File: examples/StoreAndRetrieve/nodejs/loadData.js
// Function for generating random JSON documents

var person = {
  firstName: ["John", "Markus", "Gerald", "Adrian", "Dominik",
                   "Andrew", "David", "Edward", "Martin", "Penny"],
  lastName: ["Nangle", "Gratzer", "Venzl", "Richmond", "Ashley",
                   "Bond", "Smith", "Stadler", "Avril", "Jones"],
  age: [29, 43, 48, 42, 21, 32, 56]
};

function generateJSON() {
  var data = {
    firstName: person.firstName[Math.floor(Math.random()*person.firstName.length)],
    lastName: person.lastName[Math.floor(Math.random()*person.lastName.length)],
    age: person.age[Math.floor(Math.random()*person.age.length)]
  };
  return JSON.stringify(data);
}

// Get the required Oracle driver module that I've built before
var oracledb = require("oracledb");

// Set interval to execute code every 2 seconds
setInterval(function () {
  // Open a connection to the database
  oracledb.getConnection({
      // The connection details to my database
      user : "TEST",
      password : "test",
      connectString : "gvenzl-virtual:1521/TEST"
    },
    // Callback function holding our connection
    function(err, connection) {
      if (err) {
        console.error(err.message);
        return;
      }
      // Execute an insert into the database
      connection.execute("INSERT INTO JSON VALUES (:jsonDoc)",
        // Bind jsonDoc to the return value of gererateJSON() function
        {jsonDoc: generateJSON()},
        // Callback function holding our result
        function(err, result) {
          if (err) {
            console.error(err.message);
            return;
          }
          // Commit data to the database
          connection.commit(function (err) {
            if (err) {
              console.error(err.message);
              return;
            }
          });
          // Get amount of inserted rows. This will be 1 but gives a nice example of how to use rowsAffected
          console.log(result.rowsAffected + " row inserted.");
      });
  });
}, 2000);

File: examples/StoreAndRetrieve/python/LoadData.py
import json
import random
import time
import cx_Oracle

# Helper variables holding some example data to generate JSON from
firstName = [ "John", "Markus", "Gerald", "Adrian", "Dominik", "Andrew", "David", "Edward", "Martin", "Penny" ]
lastName = [ "Nangle", "Gratzer", "Venzl", "Richmond", "Ashley", "Bond", "Smith", "Stadler", "Avril", "Jones" ]
age = [ 29, 43, 48, 42, 21, 32, 56 ]

def generateJSON():
    """Generate random JSON document"""
    return json.dumps({ "firstName": firstName[random.randint(0, len(firstName)-1)],  
                        "lastName": lastName[random.randint(0, len(lastName)-1)],
                        "age": age[random.randint(0,len(age)-1)] })

# Main try clause (to gracefully catch Ctrl+C and close DB connection)
try:
    # Open Oracle Database connection
    dbConn = cx_Oracle.connect('test/test@//gvenzl-virtual:1521/TEST')
    
    # Open and prepare cursor
    cur = dbConn.cursor()
    cur.prepare("INSERT INTO JSON VALUES (:input)")
 
    # Endless loop until user hits Ctrl+C
    while True:
        # Set "input" directory bind variable to generated JSON
        input = [ (generateJSON()) ]
        # Execute the statement. The cx_Oracle documentation says that it won't re-prepare the statement if it is the same as before. So technically I could just post the insert statement directly and don't prepare the cursor at all. However, for educational purposes it is better to show how to prepare a cursor and use None to make use of the prepared cursor.
        cur.execute(None, input)
        # Commit changes to the database
        dbConn.commit()
        print (str(cur.rowcount) + " row(s) inserted")

        # Sleep for 2 seconds
        time.sleep(2)
# Capture Ctrl+C and SystemExit and close the dataase connection
except (KeyboardInterrupt, SystemExit):
    print ("Closing Database connection")
    cur.close()
    dbConn.close()
    print ("Stopping program")
